ClassCircularityError when using the agent with Tomcat

Even loading the agent as soon as the web application is deployed, I am hitting a ClassCircularityError in ConcurrentHashMap.

<img width="1126" alt="Screen Shot 2022-12-02 at 17 18 13" src="https://user-images.githubusercontent.com/1818187/205258302-f3284b36-be6c-41e4-b3b6-eb6d507cdb0f.png">


## Expected Behavior
Agent loads fine

## Actual Behavior
The above ClassCircularityError causes the following additional issues which prevent from using the agent at all

### error cannot even be outputted due to 
```
failed to access class reactor.blockhound.BlockHoundRuntime from class reactor.blockhound.integration.StandardOutputIntegration$PrintStreamDelegate (reactor.blockhound.BlockHoundRuntime and reactor.blockhound.integration.StandardOutputIntegration$PrintStreamDelegate are in unnamed module of loader org.apache.catalina.loader.ParallelWebappClassLoader @169bd756)
```

### all threads die
With the above `IllegalAccessError`, e.g.
```Exception: java.lang.IllegalAccessError thrown from the UncaughtExceptionHandler in thread "http-nio-20088-AsyncTimeout"```

## Steps to Reproduce
Initialize the agent in a servlet

## Possible Solution
Maybe upgrade to latest bytebuddy?

## Your Environment
* Reactor version(s) used: none yet
* Other relevant libraries versions (eg. `netty`, ...): Tomcat 8.5
* JVM version (`java -version`): 11
* OS and version (eg `uname -a`): Mac OS X

Hi @mauhiz ,

Thanks for the reproducer demo.

After some struggling with the cargo documentation, maybe the following will work (actually, blockhound needs to be loaded as a shared dependency, see https://codehaus-cargo.github.io/cargo/Application+Classpath.html).

So, first, can you modify in your sample code the BlockHoundInitializer, we will use reactor-core API in order to validate if blockhound is reporting blocking code as expected:

```java
package net.mauhiz;

import reactor.core.publisher.Mono;

import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;
import javax.servlet.annotation.WebListener;
import java.time.Duration;

import reactor.blockhound.BlockHound;

@WebListener
public class BlockHoundInitializer implements ServletContextListener {

    @Override
    public void contextInitialized(ServletContextEvent sce) {
        System.out.println("XXXX Installing BlockHound");
        BlockHound.install();

        Mono.delay(Duration.ofSeconds(1))
                .doOnNext(it -> {
                    try {
                        System.out.println("XXXX sleeping");
                        Thread.sleep(10);
                        System.out.println("XXXX should not be displayed");
                    }
                    catch (InterruptedException e) {
                        throw new RuntimeException(e);
                    }
                })
                .block();
    }
}
````

next, can you modify your pom.xml like this:

```xml
    <build>
        <plugins>
            <plugin>
                <groupId>org.codehaus.cargo</groupId>
                <artifactId>cargo-maven3-plugin</artifactId>
                <version>1.10.4</version>
                <configuration>
                    <container>
                        <containerId>tomcat9x</containerId>
                        <type>embedded</type>
                        <dependencies>
                            <dependency>
                                <groupId>io.projectreactor.tools</groupId>
                                <artifactId>blockhound</artifactId>
                                <classpath>shared</classpath>
                            </dependency>
                        </dependencies>
                    </container>
                    <deployables>
                        <deployable>
                            <type>war</type>
                            <location>${project.build.directory}/${project.build.finalName}.war</location>
                            <properties>
                                <context>/</context>
                            </properties>
                        </deployable>
                    </deployables>
                </configuration>
            </plugin>
        </plugins>
    </build>

    <dependencies>
        <dependency>
            <groupId>javax.servlet</groupId>
            <artifactId>javax.servlet-api</artifactId>
            <version>4.0.1</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>io.projectreactor</groupId>
            <artifactId>reactor-core</artifactId>
            <version>3.4.26</version>
        </dependency>
        <dependency>
            <groupId>io.projectreactor.tools</groupId>
            <artifactId>blockhound</artifactId>
            <version>1.0.6.RELEASE</version>
            <scope>provided</scope>
        </dependency>
    </dependencies>
```

when running `mvn clean package cargo:run`, I'm then getting the expected error:

```
reactor.core.Exceptions$ReactiveException: reactor.blockhound.BlockingOperationError: Blocking call! java.lang.Thread.sleep
	at reactor.core.Exceptions.propagate(Exceptions.java:396)
	at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:97)
	at reactor.core.publisher.Mono.block(Mono.java:1742)
	at net.mauhiz.BlockHoundInitializer.contextInitialized(BlockHoundInitializer.java:31)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4764)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5222)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:726)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:698)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:696)
	at org.apache.catalina.startup.Tomcat.addWebapp(Tomcat.java:783)
	at org.apache.catalina.startup.Tomcat.addWebapp(Tomcat.java:733)
	at org.apache.catalina.startup.Tomcat.addWebapp(Tomcat.java:249)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.cargo.container.tomcat.internal.TomcatEmbedded$Wrapper.invoke(TomcatEmbedded.java:454)
	at org.codehaus.cargo.container.tomcat.internal.TomcatEmbedded$Wrapper.invoke(TomcatEmbedded.java:413)
	at org.codehaus.cargo.container.tomcat.internal.TomcatEmbedded$Embedded.createContext(TomcatEmbedded.java:712)
	at org.codehaus.cargo.container.tomcat.TomcatEmbeddedLocalDeployer.deploy(TomcatEmbeddedLocalDeployer.java:91)
	at org.codehaus.cargo.container.tomcat.TomcatEmbeddedLocalDeployer.redeploy(TomcatEmbeddedLocalDeployer.java:149)
	at org.codehaus.cargo.container.tomcat.internal.AbstractCatalinaEmbeddedLocalContainer.doStart(AbstractCatalinaEmbeddedLocalContainer.java:166)
	at org.codehaus.cargo.container.spi.AbstractEmbeddedLocalContainer.startInternal(AbstractEmbeddedLocalContainer.java:78)
	at org.codehaus.cargo.container.spi.AbstractLocalContainer.start(AbstractLocalContainer.java:230)
	at org.codehaus.cargo.maven3.ContainerStartMojo.executeLocalContainerAction(ContainerStartMojo.java:83)
	at org.codehaus.cargo.maven3.ContainerRunMojo.doExecute(ContainerRunMojo.java:95)
	at org.codehaus.cargo.maven3.AbstractCargoMojo.execute(AbstractCargoMojo.java:411)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:972)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:293)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:196)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
		... 48 more
Caused by: reactor.blockhound.BlockingOperationError: Blocking call! java.lang.Thread.sleep
	at java.lang.Thread.sleep(Thread.java)
	at net.mauhiz.BlockHoundInitializer.lambda$contextInitialized$0(BlockHoundInitializer.java:24)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:185)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

let me know if this works for you.
if all is working fine,  I'll update the documentation regarding the usage of Cargo + Tomcat.

Notice that we will probably make a release soon for the support of JDK 17+

thanks.

--- a/docs/quick_start.md
+++ b/docs/quick_start.md
@@ -54,6 +54,60 @@ _Gradle_
     }
 ```
 
+## Using Tomcat
+
+When using BlockHound from a Tomcat webapp, do not embedd blockhound dependency within your webapp. Instead of that, just drop
+the blockhound jar in the Tomcat shared "lib" directory.
+If you are using `Cargo` maven plugin, this can be done using a [shared classpath](https://codehaus-cargo.github.io/cargo/Application+Classpath.html) 
+
+Here is an example using Cargo maven plugin:
+
+````xml
+    <dependencies>
+        <dependency>
+            <groupId>io.projectreactor.tools</groupId>
+            <artifactId>blockhound</artifactId>
+            <version>(latest blockhound version)</version>
+            <scope>provided</scope>
+        </dependency>
+        ...
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.codehaus.cargo</groupId>
+                <artifactId>cargo-maven3-plugin</artifactId>
+                <version>1.10.4</version>
+                <configuration>
+                    <container>
+                        <containerId>tomcat9x</containerId>
+                        <type>embedded</type>
+                        <dependencies>
+                            <dependency>
+                                <groupId>io.projectreactor.tools</groupId>
+                                <artifactId>blockhound</artifactId>
+                                <classpath>shared</classpath>
+                            </dependency>
+                        </dependencies>
+                    </container>
+                    <deployables>
+                        <deployable>
+                            <type>war</type>
+                            <location>${project.build.directory}/${project.build.finalName}.war</location>
+                            <properties>
+                                <context>/</context>
+                            </properties>
+                        </deployable>
+                    </deployables>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+    ...
+````
+
+
 ## Installation
 BlockHound is a JVM agent. You need to "install" it before it starts detecting the issues:
 ```java

